vyper/parser/parser.py

```diff
@@ -86,20 +87,36 @@ def parse_external_interfaces(external_interfaces, global_ctx):
     return external_interfaces
 
 
-def parse_other_functions(
-    o, otherfuncs, sigs, external_interfaces, global_ctx, default_function, is_contract_payable,
-):
+def parse_other_functions(o, otherfuncs, sigs, external_interfaces, global_ctx, default_function):
+    # check for payable/nonpayable external functions to optimize nonpayable assertions
+    func_types = [i._metadata["type"] for i in global_ctx._defs]
+    mutabilities = [i.mutability for i in func_types if i.visibility == FunctionVisibility.EXTERNAL]
+    has_payable = next((True for i in mutabilities if i == StateMutability.PAYABLE), False)
+    has_nonpayable = next((True for i in mutabilities if i != StateMutability.PAYABLE), False)
+    is_default_payable = (
+        default_function is not None
+        and default_function._metadata["type"].mutability == StateMutability.PAYABLE
+    )
+    # when a contract has a payable default function and at least one nonpayable
+    # external function, we must perform the nonpayable check on every function
+    check_per_function = is_default_payable and has_nonpayable
+
     # generate LLL for regular functions
+    payable_func_sub = ["seq"]
     external_func_sub = ["seq"]
     internal_func_sub = ["seq"]
     add_gas = func_init_lll().gas
 
     for func_node in otherfuncs:
+        func_type = func_node._metadata["type"]
         func_lll = parse_function(
-            func_node, {**{"self": sigs}, **external_interfaces}, global_ctx, is_contract_payable,
+            func_node, {**{"self": sigs}, **external_interfaces}, global_ctx, check_per_function
         )
-        if func_lll.context.is_internal:
+        if func_type.visibility == FunctionVisibility.INTERNAL:
             internal_func_sub.append(func_lll)
+        elif func_type.mutability == StateMutability.PAYABLE:
+            add_gas += 30
```

Comment:
random `30` lol

Reply:
this is related to the gas estimation.  we plan to remove it, but that seemed outside of the scope of this (and more work).  but yeah i agree it's weird, and hopefully will not be there.